modules/calib3d/src/usac/dls_solver.cpp

```diff
@@ -0,0 +1,881 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+
+// Copyright (C) 2019 Czech Technical University.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//
+//     * Neither the name of Czech Technical University nor the
+//       names of its contributors may be used to endorse or promote products
+//       derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Please contact the author of this library if you have any questions.
+// Author: Daniel Barath (barath.daniel@sztaki.mta.hu)
+// Modification: Maksym Ivashechkin (ivashmak@cmp.felk.cvut.cz)
+
+#include "../precomp.hpp"
+#include "../usac.hpp"
+#if defined(HAVE_EIGEN)
+#include <Eigen/Eigen>
+#elif defined(HAVE_LAPACK)
+// #include <lapacke.h>
+#include "opencv_lapack.h"
+#endif
+
+namespace cv { namespace usac {
+// This is the estimator class for estimating a homography matrix between two images. A model estimation method and error calculation method are implemented
+class DLSPnPImpl : public DLSPnP {
```

Comment:
Wondering if it is planned to use this DLS implementation for `solvePnP`and related functions?

Because currently, when using `SOLVEPNP_DLS` flags, the OpenCV code will silently use EPnP instead:
- [Ignoring the flags of solvePnP and silently executing EPNP #5065](https://github.com/opencv/opencv/issues/5065)
- [see also](https://github.com/opencv/opencv/pull/6864#issuecomment-473450679)
- unfortunately, printing an error message seems to not be an option, so adding a LOG entry?

More than 5 years later, there are less than 5 complaints I think. Somehow.

I would still update the doc to explicitly mention that you cannot use `SOLVEPNP_UPNP` and `SOLVEPNP_DLS`.
Because for now, there is just a small note in the doc. Since nobody will carefully read the doc, I would update the doc:

![image](https://user-images.githubusercontent.com/8229425/89508355-1b54d280-d7ce-11ea-9532-c6c5e4bf95f6.png)

to add a bold mention that you cannot use `SOLVEPNP_UPNP` and `SOLVEPNP_DLS`. Or better remove these entries to not advertise these methods.

Similarly:

![image](https://user-images.githubusercontent.com/8229425/89508511-4ccd9e00-d7ce-11ea-9d2f-931a9045ba1b.png)

Add a bold mention that you cannot use `SOLVEPNP_UPNP` and `SOLVEPNP_DLS`.

Reply:
Also, wondering if it is the ["improved" DLS version](http://heschian.io/index.php?page=software) that is implemented? 

> 05/17/12 - Added "robust" dls_pnp version which avoids the singular configuration of the Cayley paremeters by solving the system 3 times (with the original system centered, and rotated about the x/y/z axes, respectively).

Reply:
I would say that I planned to use this DLS solver (Joel A. Hesch and Stergios I. Roumeliotis) inside my RANSAC. This solver will update so-far-the-best model by non-minimal sample.

Reply:
(Not related to this PR).

I believe once this work will be merged, I will try to propose a PR to reactivate the `SOLVEPNP_DLS` code path in `solvePnP`.

For now, I think this DLS needs either OpenCV built with LAPACK or Eigen support. So the "reactivation" could only take place if OpenCV embeds LAPACK.